### PR TITLE
More consistent with/add methods for both Scala and Java APIs

### LINKED ISSUE #83: Add withCookies to WSRequest
The WsCookie support could be improved.

```
def obtainCookie = ws.url(myurl).post(credentials).map {
      response => response.cookies
 }
```

returns a Seq[WsCookie]'s but there is no WsCookie support in the builder:

```
ws.url("request").withCookies(obtainCookies)
```

### PR DESCRIPTION
**Pull request by marcospereira**
## Fixes

Fixes #83 and #97.

## Purpose

This adds the ability to manipulate cookies and other request attributes using a more uniform and consistent API.

1. `with*` methods were deprecated since they don't have a clear semantic when also compared with other Play APIs. The current behavior was kept.
2. `add*` methods were added and are supposed to append values to attributes, preserving existent values
3. `set*` methods were added and are supposed to completely replace the existent values, discarding them.

More tests were added (or just rearranged) to better ensure the behavior described above.

## Background Context

See discussion here: https://github.com/playframework/playframework/issues/7016

### CODE REVIEW

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java [hunk 1]
```diff
@@ -20,31 +20,43 @@ public AhcWSCookie(play.shaded.ahc.org.asynchttpclient.cookie.Cookie ahcCookie)
     /**
      * Returns the underlying "native" object for the cookie.
      */
-    public Object getUnderlying() {
+    @Override
+    public play.shaded.ahc.org.asynchttpclient.cookie.Cookie getUnderlying() {
```
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java]: So we don't need to cast both the `WSCookie` to `AhcWSCookie` and the `getUnderlying` value.

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java [hunk 2]
```diff
@@ -20,31 +20,43 @@ public AhcWSCookie(play.shaded.ahc.org.asynchttpclient.cookie.Cookie ahcCookie)
...
         return ahcCookie.getMaxAge();
     }
 
+    @Override
     public boolean isSecure() {
         return ahcCookie.isSecure();
     }
+
+    @Override
+    public boolean httpOnly() {
```
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java]: This was missing maybe by accident?

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java [hunk 3]
```diff
@@ -32,7 +32,9 @@
      *
      * @return a promise to the response
      */
-    CompletionStage<? extends StandaloneWSResponse> get();
+    default CompletionStage<? extends StandaloneWSResponse> get() {
```
Comment (marcospereira) [inline play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java]: Using default methods here will make this interface more simple and direct to implement.

The only doubt here is if `StandaloneAhcWSRequest` should instead return a `CompletionStage<StandaloneAhcWSResponse>`. It was currently return just like declared here in the interface.

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java [hunk 4]
```diff
@@ -44,31 +46,52 @@
...
+        return setMethod("PATCH").setBody(body).execute();
+    }
+
+    /**
+     * Perform a PATCH on the request asynchronously.
+     *
+     * @param body the body of request
+     * @return a promise to the response
+     */
+    default <U> CompletionStage<? extends StandaloneWSResponse> patch(Source<ByteString, U> body) {
```
Comment (marcospereira) [inline play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java]: These `patch`, `post` and `put` methods for `Source` were added so we can keep this consistent with `setBody` versions.

**Comment by gmethvin** [comment]
I'm really against using `set` for any API that doesn't actually mutate the object. Using the `set` naming is for mutators is so ubiquitous that I'm afraid it would cause more confusion. If I have a `set*` method, I really expect to be able to do this:

```scala
val foo = new Foo()
foo.setBar(bars)
foo // foo now has a bar
```

Looking at the Akka HTTP model, it seems like they have settled on using `withHeaders` for creating a copy with a completely replaced list of headers and `addHeader`/`addHeaders` for creating a copy with an additional header: http://doc.akka.io/japi/akka-http/10.0.5/akka/http/scaladsl/model/HttpMessage.html.

Are there any other Scala APIs we can look at as an example of how this kind of method naming is done? I'm reluctant to force API changes without a strong convention behind it.

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java (hunk 1 above)
Comment (wsargent) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java]: I'm going to refactor `WSCookie` in any case as part of https://github.com/playframework/play-ws/issues/97

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java (hunk 2 above)
Comment (wsargent) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java]: Part of https://github.com/playframework/play-ws/issues/97

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala [hunk 5]
```diff
@@ -44,8 +45,8 @@ case class StandaloneAhcWSRequest(
   override type Self = StandaloneWSRequest
   override type Response = StandaloneWSResponse
 
-  require(client != null, "Null client!")
-  require(url != null, "Null url!")
+  require(client != null, "A [[StandaloneAhcWSClient]] is required, but it is null")
```
Comment (wsargent) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: The [[StandaloneAhcWSClient]] reference should be escaped

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java [hunk 6]
```diff
@@ -268,8 +367,61 @@
...
+    /**
+     * Sets the query string parameters. This will discard existent values.
+     *
+     * @param params the query string parameters
+     * @return the modified WSRequest.
+     */
+    StandaloneWSRequest setQueryString(Map<String, List<String>> params);
+
+    /**
+     * Add a new cookie. This can be called repeatedly and will preserve existent cookies.
```
Comment (wsargent) [inline play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java]: s/existent/existing/

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala [hunk 7]
```diff
@@ -65,18 +67,31 @@ case class StandaloneAhcWSRequest(
   override def withAuth(username: String, password: String, scheme: WSAuthScheme): Self =
     copy(auth = Some((username, password, scheme)))
 
-  override def withHeaders(hdrs: (String, String)*): Self = {
-    val headers = hdrs.foldLeft(this.headers)((m, hdr) =>
-      if (m.contains(hdr._1)) m.updated(hdr._1, m(hdr._1) :+ hdr._2)
-      else m + (hdr._1 -> Seq(hdr._2)))
-    copy(headers = headers)
+  override def setHeaders(hdrs: (String, String)*): Self = {
+    var newHeaders = hdrs.foldLeft(TreeMap[String, Seq[String]]()(CaseInsensitiveOrdered)) {
```
Comment (gmethvin) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: See #48. I think we should switch to a list for both query params and headers. The order does actually matter for some headers with the same name, and this will fix that bug.

**Comment by gmethvin** [comment]
I do recognize that we already have a `withHeaders` method with the "wrong" behavior, so we need to deprecate it. I just don't think `set` should be the standard convention across the API.

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala (hunk 7 above)
Comment (gmethvin) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: And if the main representation is a map, this should allow to set a map.

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StandaloneAhcWSRequest.java [hunk 8]
```diff
@@ -93,24 +91,19 @@ public StandaloneAhcWSRequest setRequestFilter(WSRequestFilter filter) {
...
         return this;
     }
 
-    /**
-     * Sets a query string
-     *
-     * @param query the query string
-     */
+    @Override
+    public StandaloneAhcWSRequest setHeaders(Map<String, List<String>> headers) {
```
Comment (hepin1989) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StandaloneAhcWSRequest.java]: plus `setHeader(String header,String value)`?

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java (hunk 2 above)
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/AhcWSCookie.java]: Done. Also added the same method for Scala API.

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java (hunk 6 above)
Comment (marcospereira) [inline play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java]: Done.

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala (hunk 7 above)
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: Can we do that in another PR? Just to avoid have two discussions here. :-)

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StandaloneAhcWSRequest.java (hunk 8 above)
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StandaloneAhcWSRequest.java]: I don't get your point. `setHeader` is implemented as a default method at the interface.

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala (hunk 5 above)
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: What do you mean by "escaped"? Not using ``[[]]``?

**Comment by marcospereira** [comment]
> I do recognize that we already have a withHeaders method with the "wrong" behavior, so we need to deprecate it. I just don't think set should be the standard convention across the API.

Agree with that. The whole point of this PR is move the APIs to a more consistent direction. If we could find a path that also help us to be consistent with other popular Scala APIs, that would be even better.

I think we have two options here:

1. Find another naming scheme and drop `set`.
2. Accept we need to break the "wrong" behavior of `with` methods and proper document it in our migration guides.

But giving it a second thought, I'm considering the following: do we really need to have `with` and `add` methods, or should we have just `with` methods and always consider it as an appending operation? I mean, it there the need to have a method to completely reset/replace the headers/query-string parameters/etc? I couldn't think about use cases where people need to discard existing values and set new ones.

WDYT?

**Comment by marcospereira** [comment]
> Looking at the Akka HTTP model, it seems like they have settled on using withHeaders for creating a copy with a completely replaced list of headers and addHeader/addHeaders for creating a copy with an additional header: http://doc.akka.io/japi/akka-http/10.0.5/akka/http/scaladsl/model/HttpMessage.html.

Ok, so let's keep consistent with that. :-)

By the way, we can kind of keep the `with` names and deprecate the old ones here:

1. `withHeaders` would be deprecated because if appends values
2. `withHttpHeaders` is the more verbose name with the right behavior (replaces the existing value) and also the right name.

WDYT?

**Comment by gmethvin** [comment]
@marcospereira ok, that sounds reasonable. For the Java API I think we could either change the API to be immutable, or switch to the names you suggest here.

**Comment by schmitch** [comment]
I would prefer to actually make it immutable.
It's way more simple and I think the overhead isn't too big that it's too bad.

Having it immutable also makes it easier to actually implement the Java API with the Scala API.

**Comment by gmethvin** [comment]
Yeah, I agree @schmitch. There is a migration cost for users when we make API changes, so if we think immutable makes sense we should do it now.

For the most part the rest of the API is immutable, though `WSResponse#getAllHeaders` returns a mutable map.

**Comment by marcospereira** [comment]
Methods were renamed as stated here: https://github.com/playframework/play-ws/pull/102#issuecomment-297489979. The Java methods kept the `set` pattern because that is what the whole API is using.

I'm not planning in solving any mutability issues here since this PR is already has a lot of API changes. We can handle that in another PR later.

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala (hunk 5 above)
Comment (wsargent) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: "A StandaloneAhcWSClient is required, but it is null" the [[ ]] reference is for Scaladoc link, which isn't processed in the string here -- it goes into an exception.

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java (hunk 3 above)
Comment (wsargent) [inline play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSRequest.java]: It has to be this way because of variance -- if you're extending StandaloneWSRequest with WSRequest (which adds Play-specific types) then you have to specify that the return type is covariant.

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala (hunk 5 above)
Comment (marcospereira) [inline play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSRequest.scala]: Done!
